cql3/statements/create_keyspace_statement.cc

```diff
@@ -148,34 +148,37 @@ std::optional<sstring> check_restricted_replication_strategy(
+            {"org.apache.cassandra.locator.LocalStrategy", locator::replication_strategy_type::local}
+    };
+
     sstring replication_strategy = locator::abstract_replication_strategy::to_qualified_class_name(
         *attrs.get_replication_strategy_class());
-    auto& topology = qp.proxy().get_token_metadata_ptr()->get_topology();
-    // SimpleStrategy is not recommended in any setup which already has - or
-    // may have in the future - multiple racks or DCs. So depending on how
-    // protective we are configured, let's prevent it or allow with a warning:
-    if (replication_strategy == "org.apache.cassandra.locator.SimpleStrategy") {
-        switch(qp.db().get_config().restrict_replication_simplestrategy()) {
-        case db::tri_mode_restriction_t::mode::TRUE:
-            throw exceptions::configuration_exception(
-                "SimpleStrategy replication class is not recommended, and "
-                "forbidden by the current configuration. Please use "
-                "NetworkToplogyStrategy instead. You may also override this "
-                "restriction with the restrict_replication_simplestrategy=false "
-                "configuration option.");
-        case db::tri_mode_restriction_t::mode::WARN:
-            return format("SimpleStrategy replication class is not "
-                "recommended, but was used for keyspace {}. The "
-                "restrict_replication_simplestrategy configuration option "
-                "can be changed to silence this warning or make it into an error.",
-                keyspace);
-        case db::tri_mode_restriction_t::mode::FALSE:
-            // Scylla was configured to allow SimpleStrategy, but let's warn
-            // if it's used on a cluster which *already* has multiple DCs:
-            if (topology.get_datacenter_endpoints().size() > 1) {
-                return "Using SimpleStrategy in a multi-datacenter environment is not recommended.";
-            }
-            break;
+    if (const auto& strategy_mapping = rs_to_type.find(replication_strategy); strategy_mapping != rs_to_type.end()) {
+        auto selected_strategy = strategy_mapping->second;
+        auto rs_warn_list = qp.db().get_config().replication_strategy_warn_list();
+        auto rs_fail_list = qp.db().get_config().replication_strategy_fail_list();
+
+        if (auto present_on_fail_list = std::find(rs_fail_list.begin(), rs_fail_list.end(), selected_strategy); present_on_fail_list != rs_fail_list.end()) {
+            ++stats.replication_strategy_warn_list_violations;
+            throw exceptions::configuration_exception(format(
+                    "{} replication class is not recommended, "
+                    "and forbidden by the current configuration, but was used for keyspace {}. "
+                    "Consider using e.g. NetworkTopologyStrategy instead. You may also override this "
+                    "restriction by modifying replication_strategy_fail_list configuration option.",
```

Comment:
I think it would have been nicer to say exactly how it needs to be modified. For example:
"SimpleStrategy is forbidden by the current configuration .... You may also override this restriction by setting the replication_strategy_fail_list configuration option to not list SimpleStrategy" (or "simple", if that's what you require in the configuration).

Reply:
I thought the comment in scylla.yaml, to which we are redirecting here, already explains that (in the current implementation, which may change some day) it's enough to remove given strategy from a specific list. So basically this message I would keep generic and provide more tech details around the definition of this option, so scylla.yaml, which - I believe - is already the case (correct me if I'm wrong).

Reply:
First of all, scylla.yaml isn't the only way to set configuration, so comments there don't make it redundant to explain things properly in other places.
As I noted above, it's a problem that the error message says "NetworkTopologyStrategy is not recommend and forbidden.... etc." but to allow it you need to guess that it is the word "network" that you need to remove from "replica_strategy_fail_list".
Again my overarching "rule" is that if the error message tries to tell you how to fix the error, the explanation should be accurate and as complete as possible - not contain wrong information (like in my other comment about NetworkToplogyStrategy) and not partial information that only hints what needs to be done.

Reply:
Provided more accurate description.